portfolio/src/main/java/com/google/sps/servlets/ChooseUsernameServlet.java

```diff
@@ -34,21 +39,51 @@
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     UserService userService = UserServiceFactory.getUserService();
     if (!userService.isUserLoggedIn()) {
-      response.sendRedirect("/index.html");
+      response.setStatus(HttpServletResponse.SC_UNAUTHORIZED,
+          "You cannot set a username if you are not logged in.");
       return;
     }
 
     String username = request.getParameter("username");
     String id = userService.getCurrentUser().getUserId();
 
+    // Build transaction with cross-group functionality
+    // so it can put both a Username and UserInfo entity
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    // Make a UserInfo Entity with retrieved ID and username
-    Entity entity = new Entity("UserInfo", id);
-    entity.setProperty("id", id);
-    entity.setProperty("username", username);
-    // The put() function automatically inserts new data or updates existing data based on ID
-    datastore.put(entity);
+    TransactionOptions options = TransactionOptions.Builder.withXG(true);
+    Transaction transaction = datastore.beginTransaction(options);
+    try {
+      Entity usr = new Entity("Username", username);
+      Entity check = null;
+      try {
+        Key usrkey = KeyFactory.createKey("Username", username);
+        check = datastore.get(usrkey);
+      } catch (EntityNotFoundException e) {
+        // This is what we want (the entity was not found, free to use the username)
+      }
 
+      if (check != null) {
+        // An entity was found with the requested username
+        response.setStatus(409, "Username already exists.");
+        response.getWriter().close();
```

Comment:
It's usually good to close writers you create. However, in this case, I don't think you actually created the writer. As such, you should rely on the framework to close it for you. The general rule of thumb is, "whoever opens/owns the opened resource should also be closing it."

In Java, there's a "try-with-resource" semantic, which is syntactic sugar to make it so stuff you open is always closed:

```
try (SomeWriter w = openSomeWriter(<filename or whatever>)) {
    // Do stuff...
    w.write("blah blah");
}
```

In this case, since you didn't open the writer, it's best to just leave it alone, UNLESS the servlet documentation tells you that you need to close it.

What I've found about it: https://stackoverflow.com/questions/1159168/should-one-call-close-on-httpservletresponse-getoutputstream-getwriter

And it looks like the

Reply:
Ah yes, that makes a lot of sense. I think I had the closing writer and committing transaction left in there from when I thought things going through because of that, when it turned out to be an issue with my javascript async function. Took them out and it works the same and makes much more sense! Thank you!